api/maven-api-core/src/main/java/org/apache/maven/api/DependencyProperties.java

```diff
@@ -33,27 +36,226 @@
+         *
+         * @see #intern()
+         */
+        private static final Map<String, Key<?>> INTERNS = new HashMap<>();
+
+        /**
+         * Value returned by {@link #name()}.
+         */
+        @Nonnull
+        private final String name;
+
+        /**
+         * Value returned by {@link #valueType()}.
+         */
+        @Nonnull
+        private final Class<V> valueType;
+
+        /**
+         * Creates a new key.
+         *
+         * @param name name of the key
+         * @param valueType type of value associated to the key
+         */
+        public Key(@Nonnull final String name, @Nonnull final Class<V> valueType) {
+            this.name = Objects.requireNonNull(name);
+            this.valueType = Objects.requireNonNull(valueType);
+        }
+
+        /**
+         * If a key exists in the {@linkplain #intern() intern pool} for the given name, returns that key.
+         * Otherwise, if the {@code defaultType} is non-null, creates a key for values of the specified type.
+         * Otherwise, returns {@code null}.
+         *
+         * @param name name of the key to search or create
+         * @param defaultType value type of the key to create if none exist for the given name, or {@code null}
+         * @return key found or created, or {@code null} if no key was found and {@code defaultType} is null
+         *
+         * @see #intern()
+         */
+        public static Key<?> forName(String name, Class<?> defaultType) {
+            Key<?> key;
+            synchronized (INTERNS) {
+                key = INTERNS.get(name);
+            }
+            if (key == null && defaultType != null) {
+                key = new Key<>(name, defaultType);
+            }
+            return key;
+        }
+
+        /**
+         * {@return the name of the key}.
+         */
+        @Nonnull
+        public String name() {
+            return name;
+        }
+
+        /**
+         * {@return the type of value associated to the key}.
+         */
+        @Nonnull
+        public Class<V> valueType() {
+            return valueType;
+        }
+
+        /**
+         * {@return a canonical representation of this key}. A pool of keys, initially empty, is maintained privately.
+         * When the {@code intern()} method is invoked, if the pool already contains a key equal to this {@code Key}
+         * as determined by the {@link #equals(Object)} method, then the key from the pool is returned. Otherwise,
+         * if no key exist in the pool for this key {@linkplain #name() name}, then this {@code Key} object is added
+         * to the pool and {@code this} is returned. Otherwise an {@link IllegalStateException} is thrown.
+         *
+         * @throws IllegalStateException if a key exists in the pool for the same name but a different class of values.
+         *
+         * @see String#intern()
+         */
+        @SuppressWarnings("unchecked")
+        public Key<V> intern() {
```

Comment:
I don't think the whole `intern` thing is needed at all.
I would suppose all references to `Key`s will be done using `DependencyProperties.XXX`, so I don't see a good use case for users to create duplicate instances.

Reply:
The `intern()` mechanism is for the `org.apache.maven.internal.impl.DefaultDependency` constructor. That constructor takes in argument an instance of `org.eclipse.aether.graph.Dependency` and copies its properties. But the Eclipse's properties are `<String, String>` entries, so the `String` keys need to be converted to `DependencyProperties.Key` instances. This is the purpose of `DependencyProperties.Key.forName(…)`, which provides `String` → `Key` conversions. The `intern()` stuff is a side-effect of the need to support that conversion. It could be removed if `DefaultDependency` wasn't merely a wrapper for `org.eclipse.aether.graph.Dependency`.

Reply:
Even if we restrict the scope to standard Java tools, the extensibility is still needed for supporting the `--patch-module` option, which can be seen as class-paths specific to each module. So a new enumeration value needs to be created (but not internalized) for each module to patch. An alternative design would be to make `JavaPathType` an `Enum` and support the `--path-module` option in a separated implementation of `PathType`. Whether it would make the API simpler or not is unclear to me.

Reply:
We need extensibility, as tomorrow somebody could come up with custom package that would need some special handling... story is same as in Mave3 and ArtifactHandlers... they have to be extensible, that is the whole point.

Reply:
Would it make sense to remove the `intern()` method and redefine `forName` as:
```
        @Nullable
        public static <T> Key<T> forName(@Nonnull String name, @Nullable Class<T> defaultType, boolean intern) {
            Key<?> key;
            synchronized (INTERNS) {
                key = INTERNS.get(name);
                if (defaultType != null) {
                    if (key == null) {
                        key = new Key<>(name, defaultType);
                        if (intern) {
                            INTERNS.put(name, key);
                        }
                    } else if (key.valueType() != defaultType && intern) {
                        throw new IllegalStateException("Key " + name + " already exists for a different class of values.");
                    }
                }
            }
            return (Key) key;
        }
```
The constructor could be made `private`, so that the only entry point is the above method.

Reply:
Yes, we can do that.

Reply:
Actually the above code does not fit exactly the `DefaultDependencies` needs, because it forces the key to be of the type specified by `defaultType`. This is the correct thing to do for above method because its return value is `Key<T>`. However, `DefaultDependencies` needs to first check for a key with whatever type an existing key may have. So we cannot escape the need to have two methods, one with return value `Key<?>` and another one with return value `Key<V>`.

I will replace the existing `forName` method with a simpler one like below, checking only for existing key:

```
Optional<Key<?>> forName(String);
```

As a replacement of `intern()`, I considered adding the following method:

```
<V> Key<V> createAndCache(String name, Class<V> valueType);
```

However it would not work if some plugin wants to define its own subclass of `Key`, while `intern()` allows this flexibility. I'm not sure if it is a flexibility worth to have however.